### ISSUE TITLE
How to handle Ordinal input and output?

### ISSUE THREAD

**Issue by gowthamnatarajan** [ray-project/ray#4639]
The observation space is ordinal, like 5, 10, 18, 23, 31 and the action is to choose any 3 from the input, like, 10, 18, 31,
How best to represent the above observation and action space?
I cannot make it a continuous space because, the action MUST be a subset of the observation space. Continuous action space can output ANY number and not just from the observation space.

I cannot make it discrete because, the input numbers can be anything. In one sample it could be 5, 10, 18, 23, 31. In the next sample it could be 7, 10, 15, 20, 51 etc.

To summarize, the input is a set of numbers and the output must be a subset of those numbers.

**Comment by ericl** [comment]
That's exactly what parametric actions can do, check it out: https://ray.readthedocs.io/en/latest/rllib-models.html#variable-length-parametric-action-spaces

**Comment by gowthamnatarajan** [comment]
I am confused. Those parametric actions can handle variable length actions. How do they make sure the action space is a subset of observation space?

**Comment by gowthamnatarajan** [comment]
Thanks. That helps a lot.
But, how to make sure, the action is a subset of the observation space? That is, the action is to choose something from the observation space.

**Comment by gowthamnatarajan** [comment]
Thanks. It helps a lot.
But, how do I make sure, the actions are a subset of the observation? For example, the observation can have any 5 numbers. The output should be a max of any 3 numbers out of those 5.

**Comment by ericl** [comment]
You can use an action mask as in the part right actions example. The mask
can be based on arbitrary parts of the input observation.

**Comment by gowthamnatarajan** [comment]
Any example with variable number of actions as output?

**Comment by ericl** [comment]
You could try MultiDiscrete() or Tuple(list of action spaces) for that. For example, the first action of the tuple could be Discrete(n) that determines the number of actions to take, which are the rest of the tuple elements.

**Comment by gowthamnatarajan** [comment]
Thanks. This would avoid an RNN.
I also have variable number of features in the input (observation space). I am planning to create enough dimentions to store maximum possible number of features and use padding with zeros when the input size is less than the max possible number of features.
Is that a good practice? Can something better be done other than an RNN?

**Comment by ericl** [comment]
Padding with zeros sounds fine. If the order of the features doesn't matter, you could also consider network elements like attention or pooling to make the processing order-agnostic. RNN could also work, I would just try all of them and see which works best.

**Comment by gowthamnatarajan** [comment]
The input is ordinal. That is, they are all numbers and can be sorted. And euclidean distance applies to the input. But the range of numbers can be different: Example: It could be any of the following:

1000, 1010, 1020, 1030, 1050
2200, 2205, 2210, 2215, 2240, 2250, 2260, 2300, 2400

So the numbers are in increasing order, and have variable length. The difference between the numbers in the sequence also vary.

**Comment by ericl** [comment]
I see, can it also be represented as a many-hot vector? I.e. start with a
array of zeros up to MAX_N and set all the present ordinal indices to 1.

**Comment by gowthamnatarajan** [comment]
It can. But the total number of possible numbers could be in the thousands. It may be too many. However the max number of numbers in a single sample is no more than 300. Will try several methods. Thanks.

The action space is to choose up to 4 numbers from the input. 
For example, for the 1st example the output could be 1000, 1010, 1030.
For the second it could be 2205, 2215, 2260, 2400.

These would actually be parametric actions. For each number it picks, it should output an integer as well.
For example: one output could be:
1000, -10
1010, +20
1030, -50

**Comment by ericl** [comment]
Ah I see. Well, if it's only a couple thousand and not tens of thousands, I would still give the vector encoding a try, since it's a pretty natural representation. You could also then represent the output as a vector or MAX_N probabilities from which to sample up to K actions per step.

```
obs: Box(0, 1, (MAX_N,))

actions:
   Tuple([
       Discrete(k),       # how many actions to take
       Simplex([1, MAX_N]),  # action probabilities
       Box(-50, 50, (MAX_N,)),  # action values
    ])
```

You also mentioned euclidean distance applies to the entries. In that case it may also make sense to the action output to be Box (DiagGaussian distribution, perhaps with dimension k), which you can round to then nearest valid input number to pick the actions. Lots of possibilities...

**Comment by gowthamnatarajan** [comment]
I think the only issue with multi-hot encoding the observation space is that the distance information might be lost. Eg: 2205 and 2215 are closer to each other than 2205 and 2400.

**Comment by gowthamnatarajan** [comment]
how to use the Simplex class? like spaces.Simples ([1,MAX_N])? or spaces.Simplex(shape=(1,MAX_N)) ?

**Comment by simon-mo** [comment]
Auto closing stale issue. @gowthamnatarajan if the solution didn't work please feel free to reopen.
